### DISCUSSION TITLE
[v6] Env Context Imports

### DISCUSSION THREAD

**Discussion by ShaMan123** [fabricjs/fabric.js#8208] [General]
The HEADER migration is almost done.
The final part is redoing the [environment context](https://github.com/fabricjs/fabric.js/blob/5ead2d6dc069385413124a03c7d4b82fc197567c/HEADER.js#L20) - browser or node.

There are 2 approaches I can think of regarding how to create the context
- as it is now, using detection and conditional imports/require
  It makes sense splitting the logic into 2 separate files and doing a conditional import.... But as far as I understand that is async. So we need to find out how to do that without breaking the entire repo. 
- `import { StaticCanvas as Canvas } from 'fabric/node'`: we can create another entry point that imports all node stuff (`IText`, `Textbox`, `Canvas`, `controls` etc. won't be imported)

**I am for the second option**

---

### Redundant JSDOM?

I wanted to understand why fabric uses JSDOM. It seems redundant to me.
`node-canvas` exposes `createCanvas`, `loadImage` and `Image` isn't that enough? 
We should expose methods instead of the generic `createElement` method and drop `fabric.document`
Perhaps the only other use is for SVG and I'm sure there's a simple way around that. Or simply add it for the sake of SVG which I know @asturur wants to be an optional import anyways.
The rest of fabric uses window/document in the browser context.

### `node-canvas` additional features
I see it supports a [PDF canvas](https://github.com/Automattic/node-canvas#pdf-output-support) as well as [SVG canvas](https://github.com/Automattic/node-canvas#pdf-output-support).
We can expose the PDF export.

---

### Detection

I have done some research regarding detection of node/browser.

https://stackoverflow.com/questions/34550890/how-to-detect-if-script-is-running-in-browser-or-in-node-js
https://stackoverflow.com/questions/17575790/environment-detection-node-js-or-browser
https://stackoverflow.com/questions/4224606/how-to-check-whether-a-script-is-running-under-node-js

#### npm packages found

https://github.com/abhirathore2006/detect-is-node - out dated, not tested
https://github.com/flexdinesh/browser-or-node - seems good, tested, lightweight, up to date

**Comment by ShaMan123** [comment]
#7451

**Comment by asturur** [comment]
I was wondering if node support can be a configuration option.
Like we have a default config for fabricJS, we could have an extra config file, not imported by default, that requires jsdom and set it up.
Window and document would be part of the configuration object or from the config could be injected in fabricJS.

Then in the node side of your app you could do:

```js
import { nodeConfig } from 'fabric/config/node'
...
...
fabric.config.configure(nodeConfig)
```

The reason why we use JSDOM is because of the dom manipulation we do with canvas and style.
Before fabricjs had a `createCanvasForNode` method, and had nodejs  specific code, but it felt good to have only a way to use it

**Comment by ShaMan123** [comment]
> The reason why we use JSDOM is because of the dom manipulation we do with canvas and style.

I think the migration allows us more radical things.
Why should the dom manipulation be part of canvas? It is needed only in a browser. Doesn't make sense in node.
We should refactor StaticCanvas and extract a base class that has all common functionality for node and browser.
Then we subclass for node with `node-canvas` specifics and we subclass for web adding dom manipulation to it.

If we are going in the direction of a different import for node why not go all the way? 
There really is no sense having `Canvas`, `IText`, `Textbox` etc. imported in node, is there?
I really like:
```js
import { NodeCanvas as Canvas } from 'fabric/node'
```
And it requires only another entry `index` file `node.ts`

**Comment by ShaMan123** [comment]
It seems that fabric can't be imported in SSR because it thinks it is running in node when imported, I saw an issue stating something about webpack and node canvas and disregarded it but I think that is the case after running the next template and getting

```bash
error - ../fabric.js/node_modules/canvas/build/Release/canvas.node
Module parse failed: Unexpected character '�' (1:2)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
(Source code omitted for this binary file)
```

**Reply by ShaMan123** [reply]
#8258

**Comment by ShaMan123** [comment]
If we make the evaluation using getters or methods supporting SSR will be possible maybe even w/o dynamic importing

**Comment by axetroy** [comment]
Consider split into monorepo

### Submodule of fabric maybe is not a good idea

This way still need to install `node-canvas` dependency even I don't need it.

even it can be a peerDependency

```js
// submodule of fabric
import { StaticCanvas as Canvas } from 'fabric/node'
```

### MonoRepo

```js
// use in node.js
import { Canvas } from 'fabric-node'

// use in Deno
import { Canvas } from 'fabric-deno'

// use in browser
import { Canvas } from 'fabric-dom'

// use in react-native
import { Canvas } from 'fabric-react-native'

// use in some special Javascript runtime. eg. https://developers.weixin.qq.com/miniprogram/en/dev/component/canvas.html
import { Canvas } from 'fabric-wechat-app'

// import fabric core
import { Fabric } from 'fabric'

new Fabric(new Canvas({}), { ...options })
```

The benifit of this way:

1. make `fabric` become platform independent ( no platform specific code, eg. browser )
2. It's easy for use in other platforms(Not just browser and Node)
3. It's easier for maintain the code

**Reply by ShaMan123** [reply]
It can be a module that you import (standalone), and we move `node-canvas` as an optional/peer dependency.
I think this means we should define an abstract class that fabric will use, then anyone can impl it and plug it into fabric and that is it (as you suggest). Things like svg canvas and pdf canvas are part of scope.

**Comment by ShaMan123** [comment]
#8632 has addressed most of this thread
